### PR TITLE
Add FreeRedis Diagnostics

### PR DESCRIPTION
**Pull request by guochen2**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues

___
### New feature or improvement
- Add FreeRedis Diagnostics

### CODE REVIEW

File: src/SkyApm.Diagnostics.FreeRedis/FreeRedisTracingDiagnosticProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+namespace SkyApm.Diagnostics.FreeRedis
+{
+
+
+    /// <summary>
+    /// FreeRedisTracingDiagnosticProcessor
+    /// </summary>
+    public class FreeRedisTracingDiagnosticProcessor : ITracingDiagnosticProcessor
+    {
+        public static readonly StringOrIntValue Free_Redis = new StringOrIntValue(7, "Redis");
```
Comment (wu-sheng) [inline src/SkyApm.Diagnostics.FreeRedis/FreeRedisTracingDiagnosticProcessor.cs:20]: Redis is a server-side ID, you should update the main repository to register a new ID with mapping

- https://github.com/apache/skywalking/blob/master/oap-server/server-starter/src/main/resources/component-libraries.yml#L135
- https://github.com/apache/skywalking/blob/96b6690e24fb306dde309922da70bc3bf16fa28d/oap-server/server-starter/src/main/resources/component-libraries.yml#L725

File: src/SkyApm.Diagnostics.FreeRedis/FreeRedisTracingDiagnosticProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            context.Span.AddTag("cache.type", "FreeRedis");
+            return context;
+        }
+
+        #region Notice
+        [DiagnosticName(FreeRedis_Notice)]
+        public void Notice([Object] NoticeEventArgs eventData)
+        {
+            var context = CreateFreeRedisLocalSegmentContext(eventData.NoticeType.ToString());
+            context.Span.AddTag("cache.statement", eventData.Log);
```
Comment (wu-sheng) [inline src/SkyApm.Diagnostics.FreeRedis/FreeRedisTracingDiagnosticProcessor.cs:57]: Please follow the virtual cache doc to update tags accordingly.

https://skywalking.apache.org/docs/main/next/en/setup/service-agent/virtual-cache/

Comment (wu-sheng) [reply src/SkyApm.Diagnostics.FreeRedis/FreeRedisTracingDiagnosticProcessor.cs:57]: `read` or `write` could be identified like this, https://github.com/apache/skywalking-java/blob/main/apm-sniffer/apm-sdk-plugin/jedis-plugins/jedis-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/jedis/v4/JedisPluginConfig.java

File: src/SkyApm.Diagnostics.FreeRedis/FreeRedisTracingDiagnosticProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            ILocalSegmentContextAccessor localSegmentContextAccessor, IConfigAccessor configAccessor)
+        {
+            _tracingContext = tracingContext;
+            _localSegmentContextAccessor = localSegmentContextAccessor;
+            _tracingConfig = configAccessor.Get<TracingConfig>();
+        }
+
+        private SegmentContext CreateFreeRedisLocalSegmentContext(string operation)
+        {
+            var context = _tracingContext.CreateLocalSegmentContext(operation);
```
Comment (liuhaoyang) [inline src/SkyApm.Diagnostics.FreeRedis/FreeRedisTracingDiagnosticProcessor.cs:45]: Redis calls should use exit segment

Comment (guochen2) [reply src/SkyApm.Diagnostics.FreeRedis/FreeRedisTracingDiagnosticProcessor.cs:45]: > Redis calls should use exit segment

I don't understand what you mean.

Comment (wu-sheng) [reply src/SkyApm.Diagnostics.FreeRedis/FreeRedisTracingDiagnosticProcessor.cs:45]: CreateExitSegmentContext.

File: src/SkyApm.Diagnostics.FreeRedis/FreeRedisTracingDiagnosticProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            string cmd = AnalysisCmd(eventData.Log);
+            context.Span.AddTag(Tags.CACHE_OP, parseOperation(cmd));
+            context.Span.AddTag(Tags.CACHE_CMD, eventData.Log);
+            if (eventData?.Exception != null)
+                context.Span.ErrorOccurred(eventData.Exception, _tracingConfig);
+            _tracingContext.Release(context);
+        }
+        #endregion
+        /// <summary>
+        /// 解析读写
```
Comment (wu-sheng) [inline src/SkyApm.Diagnostics.FreeRedis/FreeRedisTracingDiagnosticProcessor.cs:67]: Please remove CN

**Review by wu-sheng — DISMISSED** [review]
Tags are good to me. @liuhaoyang We may consider a new release.

File: src/SkyApm.Diagnostics.FreeRedis/FreeRedisPluginConfig.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
+﻿using System;
```
Comment (liuhaoyang) [inline src/SkyApm.Diagnostics.FreeRedis/FreeRedisPluginConfig.cs:1]: All code files need to add the SkyAPM copyright header. Please refer to https://github.com/SkyAPM/SkyAPM-dotnet/blob/main/src/SkyApm.Abstractions/ITracingDiagnosticProcessor.cs#L1

**Review by liuhaoyang — CHANGES_REQUESTED** [review]

**Comment by liuhaoyang** [comment]
> Tags are good to me. @liuhaoyang We may consider a new release.

+1

File: src/SkyApm.Abstractions/Common/Components.cs [hunk 6]
```diff
@@ -42,6 +42,8 @@ public static class Components
 
         public static readonly StringOrIntValue Free_SQL = new StringOrIntValue(3017, "FreeSql");
 
+        public static readonly StringOrIntValue Free_Redis = new StringOrIntValue(3018, "FreeRedis");
```
Comment (wu-sheng) [inline src/SkyApm.Abstractions/Common/Components.cs:45]: @guochen2 Please register 3018 ID in the main repository.

File: src/SkyApm.Diagnostics.FreeRedis/FreeRedisPluginConfig.cs (hunk 5 above)
Comment (guochen2) [reply src/SkyApm.Diagnostics.FreeRedis/FreeRedisPluginConfig.cs]: > All code files need to add the SkyAPM copyright header. Please refer to https://github.com/SkyAPM/SkyAPM-dotnet/blob/main/src/SkyApm.Abstractions/ITracingDiagnosticProcessor.cs#L1

ok

**Review by liuhaoyang — APPROVED** [review]

**Comment by liuhaoyang** [comment]
LGTM

**Comment by liuhaoyang** [comment]
> Tags are good to me. @liuhaoyang We may consider a new release.

Recently I will prepare for the testing and release v2.2.0

**Comment by wu-sheng** [comment]
@liuhaoyang Look like we miss updating https://github.com/SkyAPM/SkyAPM-dotnet/blob/main/docs/Supported-list.md for a while.

**Pull request merged**